<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="LUCIDus">
<title>LUCIDus: an R package to implement integrative clustering • LUCIDus</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LUCIDus: an R package to implement integrative clustering">
<meta property="og:description" content="LUCIDus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">LUCIDus</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/LUCIDus.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/USCbiostats/LUCIDus/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>LUCIDus: an R package to implement integrative clustering</h1>
                        <h4 data-toc-skip class="author">Yinqi Zhao,
David V. Conti</h4>
            
            <h4 data-toc-skip class="date">2022-11-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/USCbiostats/LUCIDus/blob/HEAD/vignettes/LUCIDus.Rmd" class="external-link"><code>vignettes/LUCIDus.Rmd</code></a></small>
      <div class="d-none name"><code>LUCIDus.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about-the-lucidus-r-package">About the LUCIDus R package<a class="anchor" aria-label="anchor" href="#about-the-lucidus-r-package"></a>
</h2>
<p>Peng et al. proposed an integrative model called Latent Unknown
Clustering Integrating omics Data (LUCID) (Peng et al. 2020). LUCID
jointly estimates latent unknown clusters in multi-omics data and relate
the latent clusters to the outcome of interest. They demonstrated
performance of LUCID through extensive simulation studies and real data
applications to highlight the integration of genomic, exposomic and
metabolomic data. LUCID provides an integrated clustering framework to
researchers, has numerous downloads (18,000 times since it was first
introduced according to <strong>dlstats</strong>), and has been applied
in several environmental epidemiological studies.</p>
<p>The <strong>LUCIDus</strong> R package implements integrative
analysis of mulit-view data using the LUCID model. The toolbox of
<strong>LUCIDus</strong> contains estimation, prediction, visualization,
model selection and inference. It also allows missingness in omics data
and adjustment for covariates.</p>
</div>
<div class="section level2">
<h2 id="introduction-to-the-lucid-model">Introduction to the LUCID model<a class="anchor" aria-label="anchor" href="#introduction-to-the-lucid-model"></a>
</h2>
<p>Genetic/environmental exposures (<span class="math inline">\(\mathbf{G}\)</span>), omics data (<span class="math inline">\(\mathbf{Z}\)</span>) and outcome (<span class="math inline">\(\mathbf{Y}\)</span>) are integrated by using a
latent cluster variable, which is illustrated by the directed acyclic
graph (DAG) below. (A screenshot from <a href="https://doi.org/10.1093/bioinformatics/btz667" class="external-link">the method paper
for LUCID</a>)</p>
<p><img src="DAG.png" width="70%"></p>
<p>Let <span class="math inline">\(\mathbf{G}\)</span> be a <span class="math inline">\(n \times p\)</span> matrix with columns
representing genetic features/environmental exposures, and rows being
the observations; <span class="math inline">\(\mathbf{Z}\)</span> be a
<span class="math inline">\(n \times m\)</span> matrix of standardized
omics data and <span class="math inline">\(\mathbf{Y}\)</span> be a
<span class="math inline">\(n\)</span>-dimensional vector of outcome.
According to the DAG, we assume that <span class="math inline">\(\mathbf{G}\)</span>, <span class="math inline">\(\mathbf{Z}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span> are linked by a categorical
latent variable <span class="math inline">\(\mathbf{X}\)</span>,
consisting of <span class="math inline">\(K\)</span> categories. The DAG
implies conditional independence. The joint likelihood of the LUCID
model is formulated as <span class="math display">\[\begin{equation}
    \begin{aligned}
    l(\mathbf{\Theta}) &amp; = \sum_{i = 1}^n\log f(\mathbf{Z}_i,
Y_i|\mathbf{G_i}; \mathbf{\Theta}) \\
        &amp; = \sum_{i = 1}^n \log \sum_{j = 1}^K f(\mathbf{Z}_i|X_i =
j; \mathbf{\Theta}_j) f(Y_i|X_i = j; \mathbf{\Theta}_j) f(X_i =
j|\mathbf{G}_i; \mathbf{\Theta}_j)
    \end{aligned}
\end{equation}\]</span> where <span class="math inline">\(\mathbf{\Theta}\)</span> is a generic notation
standing for all parameters in LUCID. Additionally, we assume <span class="math inline">\(\mathbf{X}\)</span> follows a multinomial
distribution conditioning on <span class="math inline">\(\mathbf{G}\)</span>, <span class="math inline">\(\mathbf{Z}\)</span> follows a multivariate normal
distribution conditioning on <span class="math inline">\(\mathbf{X}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span> follows a Gaussian (for
continuous outcome) or Bernoulli (for binary outcome) distribution
conditioning on <span class="math inline">\(\mathbf{X}\)</span>. As a
result, the equation above is finalized as <span class="math display">\[\begin{equation}
    \begin{aligned}
         l(\mathbf{\Theta}) = \sum_{i = 1}^n \log \sum_{j = 1}^k
S(\mathbf{G}_i; \boldsymbol{\beta}_j) \phi(\mathbf{Z}_i;
\boldsymbol{\mu}_j, \boldsymbol{\Sigma}_j)f(Y_i;\mathbf{\Theta}_j)
    \end{aligned}
\end{equation}\]</span> where <span class="math inline">\(S\)</span>
denotes the softmax function for multinomial distribution and <span class="math inline">\(\phi\)</span> denotes the probability density
function (pdf) of the multivariate normal distribution.</p>
<p>Because <span class="math inline">\(\mathbf{X}\)</span> is not
observed, we use the Expectation-Maximization (EM) algorithm to obtain
the maximum likelihood estimates (MLE) of LUCID. We denote the observed
data (<span class="math inline">\(\mathbf{G}\)</span>, <span class="math inline">\(\mathbf{Z}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span>) as <span class="math inline">\(\mathbf{D}\)</span>. In the E-step, we estimate
responsibility below: <span class="math display">\[\begin{equation}
    \begin{aligned}
        r_{ij} &amp; = P(X_i = j|\mathbf{D}, \mathbf{\Theta}) \\
            &amp; = \frac{S(\mathbf{G}_i; \boldsymbol{\beta}_j)
\phi(\mathbf{Z}_i; \boldsymbol{\mu}_j,
\boldsymbol{\Sigma}_j)f(Y_i;\mathbf{\Theta}_j)}{\sum_{j = 1}^k
S(\mathbf{G}_i; \boldsymbol{\beta}_j) \phi(\mathbf{Z}_i;
\boldsymbol{\mu}_j, \boldsymbol{\Sigma}_j)f(Y_i;\mathbf{\Theta}_j)}
    \end{aligned}
\end{equation}\]</span> The responsibility is interpreted as the
posterior probability of observation <span class="math inline">\(i\)</span> being assigned to latent cluster <span class="math inline">\(j\)</span> given the observed data <span class="math inline">\(\mathbf{D}\)</span>.</p>
<p>In the M-step, we maximize the expectation of joint likelihood below,
in terms of <span class="math inline">\(\mathbf{\Theta}\)</span>. <span class="math display">\[\begin{equation}
    \begin{aligned}
        Q(\mathbf{\Theta}) = \sum_{i = 1}^n\sum_{j = 1}^k
r_{ij}\log\frac{S(\mathbf{G}_i; \boldsymbol{\beta}_j)}{r_{ij}} + \sum_{i
= 1}^n\sum_{j = 1}^k r_{ij}\log\frac{\phi(\mathbf{Z}_i;
\boldsymbol{\mu}_j, \boldsymbol{\Sigma}_j)}{r_{ij}} + \sum_{i =
1}^n\sum_{j = 1}^k r_{ij}\log\frac{f(Y_i;
\boldsymbol{\Theta}_j)}{r_{ij}}
    \end{aligned}
\end{equation}\]</span> Details of the EM algorithm for LUDID can be
found in Peng et al. (2020).</p>
</div>
<div class="section level2">
<h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a>
</h2>
<table class="table">
<colgroup>
<col width="15%">
<col width="84%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/lucid.html">lucid()</a></code></td>
<td>The main function to fit LUCID models, specified by giving
integrated data and a distribution of the outcome. It also conducts
model selection and variable selection. It returns an objcet of class
<code>lucid</code>.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/summary_lucid.html">summary_lucid()</a></code></td>
<td>Create tables to summarize a LUCID model.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/plot_lucid.html">plot_lucid()</a></code></td>
<td>Visualize LUCID models through Sankey diagrams. It supports
user-defined color palette.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/boot_lucid.html">boot_lucid()</a></code></td>
<td>Derive confidence intervals based on bootstrap resampling.</td>
</tr>
<tr class="odd">
<td><code>pred_lucid()</code></td>
<td>Predict latent clsuter and outcome using integrated data.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/est_lucid.html">est_lucid()</a></code></td>
<td>A workhorse function to fit a LUCID model</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/tune_lucid.html">tune_lucid()</a></code></td>
<td>A workhorse function to conduct variable and model selection. It
fits LUCID models across the grid of tuning parameters, including number
of latent clusters <span class="math inline">\(K\)</span> and
regularization penalties. It returns the optimal model based on BIC</td>
</tr>
</tbody>
</table>
<p><code><a href="../reference/lucid.html">lucid()</a></code> calls <code><a href="../reference/est_lucid.html">est_lucid()</a></code> and
<code><a href="../reference/tune_lucid.html">tune_lucid()</a></code> in the backend. The two workhorse functions
are not normally called directly, but they can be useful when user wants
to look into model fitting process in more details.</p>
</div>
<div class="section level2">
<h2 id="fit-lucid-model">Fit LUCID model<a class="anchor" aria-label="anchor" href="#fit-lucid-model"></a>
</h2>
<p>We use <code><a href="../reference/lucid.html">lucid()</a></code> function to fit LUCID model.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/USCbiostats/LUCIDus" class="external-link">LUCIDus</a></span><span class="op">)</span></span>
<span><span class="co"># use simulated data</span></span>
<span><span class="va">G</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">G</span></span>
<span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">Z</span></span>
<span><span class="va">Y_normal</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">Y_normal</span></span>
<span><span class="va">Y_binary</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">Y_binary</span></span>
<span><span class="va">cov</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">Covariate</span></span>
<span></span>
<span><span class="co"># fit LUCID model with continuous outcome</span></span>
<span><span class="va">fit1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># fit LUCID model with binary outcome</span></span>
<span><span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_binary</span>, family <span class="op">=</span> <span class="st">"binary"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># fit LUCID model with covariates</span></span>
<span><span class="va">fit3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_binary</span>, CoY <span class="op">=</span> <span class="va">cov</span>, family <span class="op">=</span> <span class="st">"binary"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span>
<span><span class="va">fit4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_binary</span>, CoG <span class="op">=</span> <span class="va">cov</span>, family <span class="op">=</span> <span class="st">"binary"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span></code></pre></div>
<p>User should be aware of option <code>useY</code>. By default,
<code>useY = TRUE</code>, which means we estimate the latent cluster
using the information from exposure, omics data and outcome (referred to
as supervised LUCID). Otherwise, by setting <code>useY = FALSE</code>,
we only use exposure and omics data to estimate the latent clusters
(referred to as unsupervised LUCID).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># unsupervised lucid</span></span>
<span><span class="va">fit5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, useY <span class="op">=</span> <span class="cn">FALSE</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span></code></pre></div>
<p>LUCID allows for flexibe geometric features of the latent cluster,
including volume, shape and orientation. User can use argument
<code>modelName</code> to specify the models. By default, LUCID uses a
<code>VVV</code> model without putting any restraints on the model
shape. Other available models can be found in
<code><a href="https://mclust-org.github.io/mclust/reference/mclustModelNames.html" class="external-link">mclust::mclustModelNames</a></code> under the section of multivariate
mixture. If <code>modelName = NULL</code>, LUCID will choose the model
based on data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fit LUCID model with automatic selection on covariance models</span></span>
<span><span class="va">fit6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, modelName <span class="op">=</span> <span class="cn">NULL</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span>
<span><span class="co"># check the optimal model</span></span>
<span><span class="va">fit6</span><span class="op">$</span><span class="va">modelName</span></span>
<span></span>
<span><span class="co"># fit LUCID model with a specified covariance model</span></span>
<span><span class="va">fit7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, modelName <span class="op">=</span> <span class="st">"EII"</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span></code></pre></div>
<p>LUCID has two options for initializing the EM algorithm. The first
method (also the default method) uses <code>mclust</code> and regression
to initialize parameters. The second initialize EM algorithm by random
guess.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># initialize EM algorithm by mclust and regression</span></span>
<span><span class="va">fit8</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, init_par <span class="op">=</span> <span class="st">"mclust"</span> , seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># initialize EM algorithm by random guess</span></span>
<span><span class="va">fit9</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, init_par <span class="op">=</span> <span class="st">"random"</span> , seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="interpreting-lucid-model">Interpreting LUCID model<a class="anchor" aria-label="anchor" href="#interpreting-lucid-model"></a>
</h2>
<p><code><a href="../reference/summary_lucid.html">summary_lucid()</a></code> returns a coefficient table of LUCID.
The table is divided into 3 parts: (1) association between outcome and
latent cluster; (2) association between latent cluster and omics data;
(3) association between exposure and latent clsuter.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarize LUCID model</span></span>
<span><span class="fu"><a href="../reference/summary_lucid.html">summary_lucid</a></span><span class="op">(</span><span class="va">fit1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-lucid">Visualization of LUCID<a class="anchor" aria-label="anchor" href="#visualization-of-lucid"></a>
</h2>
<p>We use a Sankey diagram to visualize LUCID model. In the Sankey
diagram, each node represents a variable in LUCID (exposure, omics data
or outcome), each line corresponds to an association between two
variables. The color of the line indicates the direction of association
(by default, dark blue refers to positive association while light blue
refers to negative association) and the width of line indicates the
magnitude of association (large effect corresponds to wider line). User
can specify color palette of Sankey diagram.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualze lucid model via a Snakey diagram</span></span>
<span><span class="fu"><a href="../reference/plot_lucid.html">plot_lucid</a></span><span class="op">(</span><span class="va">fit1</span><span class="op">)</span></span></code></pre></div>
<p><img src="sankey.png" width="70%"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change node color</span></span>
<span><span class="fu"><a href="../reference/plot_lucid.html">plot_lucid</a></span><span class="op">(</span><span class="va">fit1</span>, G_color <span class="op">=</span> <span class="st">"yellow"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_lucid.html">plot_lucid</a></span><span class="op">(</span><span class="va">fit1</span>, Z_color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># change link color</span></span>
<span><span class="fu"><a href="../reference/plot_lucid.html">plot_lucid</a></span><span class="op">(</span><span class="va">fit1</span>, pos_link_color <span class="op">=</span> <span class="st">"red"</span>, neg_link_color <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="regularization">Regularization<a class="anchor" aria-label="anchor" href="#regularization"></a>
</h2>
<p>LUCID uses regularization to select variables. Regularization is
applied to exposure and omics data separately. For variable selection
for exposures (<span class="math inline">\(\mathbf{G}\)</span>), we
apply lasso regression, which is</p>
<p><span class="math display">\[
\beta_{lasso} = \arg \max_{\beta} \{ \sum_{i=1}^N\sum_{j=1}^K r_{ij}
\log S(X_i=j|G_i, \beta_j) - \lambda_{\beta} \sum_{j=1}^K
\sum_{l=1}^P|\beta_{jl}| \}
\]</span></p>
<p>For omics data, cluster specific means are updated by <span class="math display">\[
\mu_{j, lasso} = \arg \max_{\mu} \{ \sum_{i=1}^N\sum_{j=1}^K r_{ij} \log
\phi(Z_i=j|\mu_j, \Sigma_j) - \lambda_{\mu}
\sum_{j=1}^k\sum_{l=1}^m|\mu_{jl}| \}  
\]</span> Cluster specific variance-covariance matrices are updated via
their inverse matrices <span class="math inline">\(W =
\Sigma^{-1}\)</span>. LUCID uses the graphical lasso algorithm to
optimaize the parameter below, <span class="math display">\[
W_{j, lasso} = \arg \max_{\mu} \{ \sum_{i=1}^N\sum_{j=1}^K r_{ij} (det
(W_j) - trace(S_jW_j)) - \lambda_{ W}\sum_{j=1}^k \sum_{l,o}|w_{j;l,o}|
\}
\]</span> where <span class="math inline">\(S_j\)</span> is the
empirical estimation of variance-covariance matrix.</p>
<p>User should specify the penalties to arguments <code>Rho_G</code>
(for exposures), <code>Rho_Z_Mu</code> and <code>Rho_Z_Cov</code> (for
omics data). We recommend to choose penalties based on the ranges below:
1. <code>Rho_G</code>: 0 - 1 2. <code>Rho_Z_Mu</code>: 0 - 100 3.
<code>Rho_Z_Cov</code>: 0 - 1 However, the ranges above are empirical
values based on simulation studies. User can try different values based
on their own datasets. In practice, user can pass a vector of penalties
to <code>lucid</code> and let <code>lucid</code> decide the optimal
penalty.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use LUCID model to conduct integrated variable selection</span></span>
<span><span class="co"># select exposure</span></span>
<span><span class="va">fit10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, CoY <span class="op">=</span> <span class="cn">NULL</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span>, Rho_G <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">fit11</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, CoY <span class="op">=</span> <span class="cn">NULL</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span>, Rho_G <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.1</span>, by <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select omics data</span></span>
<span><span class="va">fit12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, CoY <span class="op">=</span> <span class="cn">NULL</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span>, Rho_Z_Mu <span class="op">=</span> <span class="fl">90</span>, Rho_Z_Cov <span class="op">=</span> <span class="fl">0.1</span>, init_par <span class="op">=</span> <span class="st">"random"</span><span class="op">)</span></span>
<span><span class="va">fit13</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, CoY <span class="op">=</span> <span class="cn">NULL</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span>, Rho_Z_Mu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">50</span>, by <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, Rho_Z_Cov <span class="op">=</span> <span class="fl">0.5</span>, init_par <span class="op">=</span> <span class="st">"random"</span>, verbose_tune <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-selection">Model selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h2>
<p>We use Bayesian Information Criteria (BIC) to choose the optimal
number of latent clusters (<span class="math inline">\(K\)</span>) for
LUCID model.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># tune lucid over a grid of K (note this function may take time to run)</span></span>
<span><span class="va">tune_lucid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, K <span class="op">=</span><span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="inference">Inference<a class="anchor" aria-label="anchor" href="#inference"></a>
</h2>
<p>LUCID uses bootstrap to derive confidence interval for parameters,
given a certain confidence interval.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># conduct bootstrap resampling</span></span>
<span><span class="va">boot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/boot_lucid.html">boot_lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, model <span class="op">=</span> <span class="va">fit1</span>, R <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># use 90% CI</span></span>
<span><span class="va">boot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/boot_lucid.html">boot_lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, model <span class="op">=</span> <span class="va">fit1</span>, R <span class="op">=</span> <span class="fl">100</span>, conf <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<p>Diagnostic plot for bootstrap samples is created by</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check distribution for bootstrap replicates of the variable of interest</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">boot1</span><span class="op">$</span><span class="va">bootstrap</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="incorporating-missing-omics-data">Incorporating missing omics data<a class="anchor" aria-label="anchor" href="#incorporating-missing-omics-data"></a>
</h2>
<p>The latest version of LUCID allows missingness in omics data. We
consider 2 missing patterns in omics data: (1) list-wise missing pattern
that only a subset of observations have measured omics data and (2)
sporadic missing pattern that missingness is completely at random. We
implement a likelihood partition for (1) and an integrated imputation
based EM algorithm for (2).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fit LUCID model with block-wise missing pattern in omics data</span></span>
<span><span class="va">Z_miss_1</span> <span class="op">&lt;-</span> <span class="va">Z</span></span>
<span><span class="va">Z_miss_1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span>, <span class="fl">0.3</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">fit14</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z_miss_1</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># fit LUCID model with sporadic missing pattern in omics data</span></span>
<span><span class="va">Z_miss_2</span> <span class="op">&lt;-</span> <span class="va">Z</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">arrayInd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Z_miss_2</span><span class="op">)</span>, <span class="fl">0.3</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Z_miss_2</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Z_miss_2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Z_miss_2</span><span class="op">[</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">fit15</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lucid.html">lucid</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, Z <span class="op">=</span> <span class="va">Z_miss_2</span>, Y <span class="op">=</span> <span class="va">Y_normal</span>, family <span class="op">=</span> <span class="st">"normal"</span>, K <span class="op">=</span> <span class="fl">2</span>, seed <span class="op">=</span> <span class="fl">1008</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># check the imputed omics dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fit15</span><span class="op">$</span><span class="va">Z</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<ul>
<li>Dr. Nikos Stratakis</li>
<li>Dr. Jesse Goodrich</li>
<li>Dr. Cheng Peng</li>
<li>Dr. Leda Chatzi and her lab memebers</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yinqi Zhao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
